examples/python/src/authors/query.py

```diff
@@ -0,0 +1,92 @@
+# Code generated by sqlc. DO NOT EDIT.
+from typing import AsyncIterator, Awaitable, Iterator, Optional, overload
+
+import sqlc_runtime as sqlc
+
+from authors import models
+
+
+CREATE_AUTHOR = """-- name: create_author :one
+INSERT INTO authors (
+          name, bio
+) VALUES (
+  $1, $2
+)
+RETURNING id, name, bio
+"""
+
+
+DELETE_AUTHOR = """-- name: delete_author :exec
+DELETE FROM authors
+WHERE id = $1
+"""
+
+
+GET_AUTHOR = """-- name: get_author :one
+SELECT id, name, bio FROM authors
+WHERE id = $1 LIMIT 1
+"""
+
+
+LIST_AUTHORS = """-- name: list_authors :many
+SELECT id, name, bio FROM authors
+ORDER BY name
+"""
+
+
+@overload
+def create_author(conn: sqlc.Connection, name: str, bio: Optional[str]) -> Optional[models.Author]:
+    pass
+
+
+@overload
+def create_author(conn: sqlc.AsyncConnection, name: str, bio: Optional[str]) -> Awaitable[Optional[models.Author]]:
+    pass
+
+
+def create_author(conn: sqlc.GenericConnection, name: str, bio: Optional[str]) -> sqlc.ReturnType[Optional[models.Author]]:
+    return conn.execute_one_model(models.Author, CREATE_AUTHOR, name, bio)
```

Comment:
I've never used or seen the `@overload` operator. How common is it? Are there other ways to expose a similar sync / async function?

Reply:
I'm not sure how common it is, I'm using it here so a single function can be used with both sync and async engines and type checkers properly understand it. It's definitely something I cooked up myself though, I haven't seen this specific pattern for sync+async support anywhere else. Without it I'd have to define separate sync and async functions, and with bare functions I definitely didn't want to just prefix (or suffix) every function with `async_`.

However, if the query functions are moved into classes it's a lot simpler to just define two classes, one with sync functions and another with async functions, no method prefix/suffix is needed since they are contained within separate classes. The generation of async (or sync) query classes can also be placed behind a config option, so users who don't care about async (or sync) support don't get that generated code.